rfcs/0096-experimental-nix-derivations.md

```diff
@@ -0,0 +1,81 @@
+---
+feature: experimental-nix-derivations
+start-date: 2021-06-25
+author: gytis-ivaskevicius
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+---
+
+# Summary
+[summary]: #summary
+
+Add a way for users to use experimental Nix features with improved user-experience
+and better communication to the user that they are using an experimental feature.
+
+# Motivation
+[motivation]: #motivation
+
+Motivation:
+- Revert of (this)[https://github.com/NixOS/nixpkgs/pull/123898].
+- Noticed quite a few peeps who were very supportive of [this](https://github.com/NixOS/nixpkgs/pull/120141) PR (which got reverted).
+- Lack of visibility for the end-users that they are using experimental features.
+
+Expected outcome:
+- Improved UX.
+- Reduced risk of users using experimental features in production environments.
+- Reduced non-bug questions/issues on common support channels.
+
+# Detailed design
+[design]: #detailed-design
+
+`nixUnstable` derivation definitions should be updated in such a way so end users would be able to overwrite features with something along the lines of this:
+```nix
+nixUnstable.override { experimentalFeatures = ["flakes"]; }
```

Comment:
> only benefit I see here ...

To be fair, I also can see other benefits that derives from the motivation section: improved UX & reduced risk of forgetting options as part of a system's statefulness.

It might be that the underlaying problem that motivates those motivation items is the current "chaos":

- `nixConfig`
- `NIX_CONFIG`
- `--myflag`
- `--extra-myflag`
- ... (put the ones I forgot here)

They all have in common to reflect contextual information (system context env context, flake context, cli context).

There does no seem to be such thing like "this is the `nix` package and here is what ot does" type of user experience.